qiskit/circuit/quantumcircuit.py

```diff
@@ -1586,12 +1586,13 @@ def decompose(
         from qiskit.converters.circuit_to_dag import circuit_to_dag
         from qiskit.converters.dag_to_circuit import dag_to_circuit
 
-        dag = circuit_to_dag(self)
+        # do not copy operations, this is done in the conversion from dag_to_circuit
+        dag = circuit_to_dag(self, copy_operations=False)
         dag = HighLevelSynthesis().run(dag)
         pass_ = Decompose(gates_to_decompose)
         for _ in range(reps):
             dag = pass_.run(dag)
-        return dag_to_circuit(dag)
+        return dag_to_circuit(dag, copy_operations=True)
```

Comment:
This idea looks completely sensible to me, except I think it makes more logical sense to put the copy in the `circuit_to_dag` call and elide it from `dag_to_circuit`, because that allows `HighLevelSynthesis` and `Decompose` to mutate the DAG instructions, were they to choose to, because we know that we'll be throwing away all our extant references to `dag` at the end of this function.

Reply:
It looks like `copy_operations=True` in line 1595 is just reaffirming the default. So really what this PR is doing is avoiding a deep(ish) copy of the circuit. This apparently does not break the test suite. But might break something somewhere... why was the copy added in the first place? 

I'm concerned about mutating objects in `self` that are referenced elsewhere. This is probably the same concern @jakelishman expressed.

Assuming that copying operations takes the same time in both conversions, I agree that it makes more sense to do it in `circuit_to_dag`. There is still the chance that `HighLevelSynthesis` tucks away a reference somewhere. But at least we know where to look for it.

In any case, the general idea is clearly a good one.

Reply:
Transpiler passes should/must not keep extant references to the objects that they act on, so we're safe from that.  They are permitted to mutate their inputs assuming ownership of them, though.